### PR TITLE
feat: make ProxyAdmin self-owned Proxy

### PR DESCRIPTION
**Pull request by smartcontracts**
**Description**
Turns the ProxyAdmin into a Proxy that owns itself. Result of this is that we can treat the ProxyAdmin as a Proxy which reduces some complexity in the migration. We can then also upgrade the ProxyAdmin on L2.

### CODE REVIEW

**Comment by tynes** [comment]
The `check-l2` script needs to be updated to check for the admin + impl slots being set at the `ProxyAdmin` predeploy account

**Comment by mslipper** [comment]
Confirming - the owner of the `ProxyAdmin` would be the multisig. The admin of the proxy in front of `ProxyAdmin` would be itself. This would let us use `ProxyAdmin` to self-upgrade by calling `ProxyAdmin` from the multisig. Is this correct?

**Comment by smartcontracts** [comment]
> Confirming - the owner of the `ProxyAdmin` would be the multisig. The admin of the proxy in front of `ProxyAdmin` would be itself. This would let us use `ProxyAdmin` to self-upgrade by calling `ProxyAdmin` from the multisig. Is this correct?

That is correct.

**Comment by mslipper** [comment]
This looks good to me. Would like a review from @tynes or @maurelian before merging.

**Comment by mslipper** [comment]
Closing in favor of an alternate proposal.

File: packages/contracts-bedrock/tasks/check-l2.ts [hunk 1]
```diff
@@ -577,6 +577,9 @@ const check = {
 
     const addressManager = await ProxyAdmin.addressManager()
     console.log(`  - addressManager: ${addressManager}`)
+
+    await checkProxy(hre, 'ProxyAdmin')
+    await assertProxy(hre, 'ProxyAdmin')
```
Comment (maurelian) [inline packages/contracts-bedrock/tasks/check-l2.ts:582]:
```suggestion
    await checkProxy(hre, 'ProxyAdmin', signer.provider)
    await assertProxy(hre, 'ProxyAdmin', signer.provider)
```

Comment (tynes) [reply packages/contracts-bedrock/tasks/check-l2.ts:582]: ^^ This is correct, will break when using a URL passed via CLI if not

File: op-chain-ops/genesis/setters.go [hunk 2]
```diff
@@ -178,6 +172,7 @@ func SetImplementations(db vm.StateDB, storage state.StorageConfig, immutable im
 			return fmt.Errorf("code not set for %s", name)
 		}
 	}
+
```
Comment (tynes) [inline op-chain-ops/genesis/setters.go:175]: nit: unnecessary diff

File: op-chain-ops/genesis/layer_two_test.go [hunk 3]
```diff
@@ -92,5 +92,5 @@ func TestBuildL2DeveloperGenesisDevAccountsFunding(t *testing.T) {
 
 	gen, err := genesis.BuildL2DeveloperGenesis(config, block)
 	require.NoError(t, err)
-	require.Equal(t, 2320, len(gen.Alloc))
+	require.Equal(t, 2321, len(gen.Alloc))
```
Comment (tynes) [inline op-chain-ops/genesis/layer_two_test.go:95]: There is now 1 more additional code address in the state

File: packages/contracts-bedrock/tasks/check-l2.ts (hunk 1 above)
Comment (tynes) [reply packages/contracts-bedrock/tasks/check-l2.ts:582]: CI failing with this error:

```
tasks/check-l2.ts(581,11): error TS2554: Expected 3 arguments, but got 2.
```

**Comment by maurelian** [comment]
Specs need a small change too: 

https://github.com/ethereum-optimism/optimism/blob/9a2ffd119e18a49e0264edcd29c6c6a6206755e1/specs/predeploys.md#L256-L257

**Review by maurelian — APPROVED** [review]

**Comment by mslipper** [comment]
Running against a datadir now. Will approve once I see that work.

File: op-chain-ops/genesis/check.go [hunk 4]
```diff
@@ -76,6 +76,10 @@ var (
 		predeploys.ProxyAdminAddr: {
 			// Slot 0x00 (0) is _owner. Requires custom check, so set to a garbage value
 			ProxyAdminOwnerSlot: common.HexToHash("0xbadbadbadbad0xbadbadbadbadbadbadbadbad0xbadbadbadbad0xbadbadbad0"),
```
[the remark this answers is not in the document]
Comment (mslipper) [reply op-chain-ops/genesis/check.go:78]: Could've sworn I'd fixed this as part of another PR - checking

[the remark this answers is not in the document]
Comment (mslipper) [reply op-chain-ops/genesis/check.go:78]: Ah, they weren't fully fixed there either. Fixing fully now.

**Comment by mslipper** [comment]
Migration against the datadir worked. Merging, and will perform an end-to-end migration tonight.

**Review by mslipper — APPROVED** [review]

**Pull request merged**